src/coreclr/gc/gc.cpp

```diff
@@ -12845,8 +12847,23 @@ void gc_heap::distribute_free_regions()
         }
     }
 #else //MULTIPLE_HEAPS
-    while (decommit_step())
+    // we want to limit the amount of decommit we do per time to indirectly
+    // limit the amount of time spent in recommit and page faults
+    // we use the elapsed time since the last GC to arrive at the desired
+    // decommit size
+    // we limit the elapsed time to 10 seconds to avoid spending too much time decommitting
+    size_t decommit_step_milliseconds = min (ephemeral_elapsed, (10*1000));
+
+    decommit_step (decommit_step_milliseconds);
+
+    // transfer any remaining regions on the decommit list back to the free list
+    for (int kind = basic_free_region; kind < count_free_region_kinds; kind++)
     {
+        if (global_regions_to_decommit[kind].get_num_free_regions() != 0)
+        {
+            gc_heap* hp = pGenGCHeap;
+            hp->free_regions[kind].transfer_regions (&global_regions_to_decommit[kind]);
```

Comment:
I suspect we might need to do something with respect to commit accounting here. Looks like we are moving some committed bytes from one bucket to another.

Reply:
Interesting - distribute_free_regions right now doesn't seem to do anything concerning commit accounting, as far as I can tell. It sounds like this is an omission we should repair?

Reply:
@cshung wouldn't that already be covered by the account in `decommit_step` which actually does the decommit?

Reply:
Thanks, @Maoni0, I didn't notice @PeterSolMS already replied a week ago.

I read the code again and I think you are right, we have moved the committed bytes from the object heap to free through `return_free_region` and discounted from free during `virtual_decommit` caused by `decommit_step`.

The change here is just moving the committed bytes from the per heap free lists (i.e. the free bucket) to the global free list (i.e. again, the free bucket), so there is no need to account for that.